clis/teliod/src/core_api.rs

```diff
@@ -0,0 +1,343 @@
+                    debug!("Unable to load identifier due to {e}. Registering ...");
+                    register_machine_with_exp_backoff(
+                        &device_identity.hw_identifier,
+                        device_identity.private_key.public(),
+                        auth_token,
+                    )
+                    .await?
+                }
+            },
+        );
+    }
+
+    let mut backoff = build_exp_backoff()?;
+    let mut retries = 0;
+    loop {
+        let status = update_machine(&device_identity, auth_token).await?;
+        match status {
+            StatusCode::OK => break,
+            StatusCode::FORBIDDEN | StatusCode::UNAUTHORIZED | StatusCode::BAD_REQUEST => {
+                // exit daemon
+                return Err(Error::UpdateMachineError(status));
+            }
+            StatusCode::NOT_FOUND => {
+                debug!("Unable to update. Registering machine ...");
+                // Retry machine update after registering. To make sure everything was successful
+                // If registering fails. Close the daemon
+                device_identity.machine_identifier = Some(
+                    register_machine_with_exp_backoff(
+                        &device_identity.hw_identifier,
+                        device_identity.private_key.public(),
+                        auth_token,
+                    )
+                    .await?,
+                );
+            }
+            _ => {
+                // Retry with exp back-off
+                let _ =
+                    retry_after_backoff_delay(&mut backoff, &mut retries, "update machine").await?;
+            }
+        }
+    }
+
+    let mut options = OpenOptions::new();
+    // User has read/write but others have none
+    options.mode(0o600);
+    let mut file = options.create(true).write(true).open(identity_path)?;
+    serde_json::to_writer(&mut file, &device_identity)?;
+    Ok(device_identity)
+}
+
+async fn fetch_identifier_with_exp_backoff(
+    auth_token: &str,
+    public_key: PublicKey,
+) -> Result<MachineIdentifier, Error> {
+    let mut retries = 0;
+    let mut backoff = build_exp_backoff()?;
+    loop {
+        match fetch_identifier_from_api(auth_token, public_key).await {
+            Ok(id) => return Ok(id),
+            Err(e) => match e {
+                Error::FetchingIdentifierError => return Err(e),
+                _ => {
+                    let _ =
+                        retry_after_backoff_delay(&mut backoff, &mut retries, "fetch identifier")
+                            .await?;
+                }
+            },
+        }
+    }
+}
+
+async fn register_machine_with_exp_backoff(
+    hw_identifier: &str,
+    public_key: PublicKey,
+    auth_token: &str,
+) -> Result<MachineIdentifier, Error> {
+    let mut retries = 0;
+    let mut backoff = build_exp_backoff()?;
+    loop {
+        if let Ok(id) = register_machine(hw_identifier, public_key, auth_token).await {
```

Comment:
Please brake the retry loop if we receive the `400 - Bad request` error `101120` `Maximum machine count reached`. Or maybe in fact any error from `Bad request` or `Invalid credentials`? Since it makes no sense to retry any amount of times as it will never work.

Reply:
Done!